Hadoop-BAM: add support for efficiently loading a subset of a BAM

tomwhite opened the pull request:
Addresses #1399

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -92,17 +93,23 @@ public ReadsSparkSource(final JavaSparkContext ctx, final ValidationStringency v
-
         }).filter(v1 -> v1 != null);
     }
 
```
There is a TODO below for the `samRecordOverlaps()` method that says "remove this method when https://github.com/broadinstitute/hellbender/issues/559 is fixed", but that ticket has been closed as resolved. Can you create a new ticket "Use `IntervalsSkipList` in `ReadSparkSource.samRecordOverlaps()`, update the docs for the method to point to that ticket, and also add a note to the `samRecordOverlaps()` that it is only used as a fallback option for formats that don't support query-by-interval natively at the Hadoop-BAM layer?

tomwhite replied:
Opened https://github.com/broadinstitute/gatk/issues/1531 for this.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -92,17 +93,23 @@ public ReadsSparkSource(final JavaSparkContext ctx, final ValidationStringency v
 
+        boolean isBam = IOUtils.isBamFileName(readFileName);
+        if (isBam && intervals != null && !intervals.isEmpty()) {
+            BAMInputFormat.setIntervals(conf, intervals);
```
Have we tested `BAMInputFormat.setIntervals()` with a large interval list like `ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/b37/Broad.human.exome.b37.interval_list.gz`? I wonder whether there might be performance issues, given that it converts all the intervals into one huge String and re-parses them later.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java:
```diff
@@ -187,6 +186,21 @@ public void testCRAMReferenceFromHDFS() throws Exception {
+        JavaRDD<GATKRead> reads = readSource.getParallelReads(NA12878_chr17_1k_BAM, null, intervals);
+
+        SamReaderFactory samReaderFactory = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT);
+        SamReader samReader = samReaderFactory.open(new File(NA12878_chr17_1k_BAM));
```
try-with-resources

tomwhite replied:
Fixed.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java:
```diff
@@ -187,6 +186,21 @@ public void testCRAMReferenceFromHDFS() throws Exception {
+        SAMRecordIterator query = samReader.query(new QueryInterval[]{new QueryInterval(seqIndex, 69010, 69040), new QueryInterval(seqIndex, 69910, 69920)}, false);
+        Assert.assertEquals(reads.count(), Iterators.size(query));
+    }
+
```
It's probably worth adding a test case with the large interval list I linked to above to make sure there are no performance problems there (should probably be checked in to `src/test/resources/large` using `git lfs`). Can use the methods in `IntervalArgumentCollection` to load the interval list.

droazen commented:
:+1: merge after addressing comments. If the test I requested is too slow to be part of our test suite, should create a ticket "Fix performance issues in `BAMInputFormat.setIntervals()` with large interval lists"

tomwhite commented:
I addressed the comments except the one about the large intervals file. I agree that this is a problem and we need to test for this case (which I opened #1532 for). It looks like `IntervalArgumentCollection` doesn't support reading from a file containing a list of intervals though, despite the javadoc for `getIntervalStrings()`. There's probably a bigger issue for handling large intervals for other tools, since there are no tests for that either.

droazen commented:
@tomwhite `IntervalArgumentCollection` does work with interval files (tried it just now) -- you just need to make sure the extension is one of `IntervalUtils.INTERVAL_FILE_EXTENSIONS` (won't work with gzipped files, need to extract those first)

tomwhite commented:
@droazen ah ok. My mistake.
